### ISSUE TITLE
SIGSEGV instead of signal handler invocation on E2K

### ISSUE THREAD

**Issue by ivmai** [ivmai/bdwgc#459]
Source: master (41d41c309)
Host: Linux/e2k (GNU/Linux 5.4.0-3.19-e8c)
Compiler: lcc:1.26.12:Jun--5-2022:e2k-v4-linux gcc (GCC) 9.3.0 compatible
How to reproduce: ./autogen.sh && ./configure --disable-shared --enable-gc-assertions && make -j check CFLAGS_EXTRA="-O0 -D NO_INCREMENTAL"
Repeat `make -j check` >30 times

gdb gctest -c core:
```
Core was generated by `./gctest'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000045d4843431a0 in __lll_lock_wait () from /lib64/libpthread.so.0
[Current thread is 1 (Thread 0x45d4978671b0 (LWP 22293))]
(gdb) bt
#0  0x000045d4843431a0 in __lll_lock_wait () from /lib64/libpthread.so.0
#1  0x000045d484327720 in pthread_mutex_lock () from /lib64/libpthread.so.0
#2  0x00000000000943a8 in GC_generic_lock (lock=0x115860 <GC_allocate_ml>) at pthread_support.c:2365
#3  0x00000000000944a0 in GC_lock () at pthread_support.c:2466
#4  0x00000000000868f0 in GC_is_thread_tsd_valid (tsd=0x45d4842ec078) at pthread_support.c:765
#5  0x000000000009e4d8 in GC_malloc_kind (bytes=16, kind=0) at thread_local_alloc.c:181
#6  0x000000000004d688 in GC_malloc_atomic (lb=16) at malloc.c:344
#7  0x0000000000013d98 in small_cons_leaf (x=1258) at tests/gctest.c:406
#8  0x0000000000014858 in ints (low=1258, up=-4500) at tests/gctest.c:510
#9  0x0000000000014890 in ints (low=1257, up=-4500) at tests/gctest.c:510
#10 0x0000000000014890 in ints (low=1256, up=-4500) at tests/gctest.c:510
...

(gdb) info thr
  Id   Target Id                         Frame
* 1    Thread 0x45d4978671b0 (LWP 22293) 0x000045d4843431a0 in __lll_lock_wait () from /lib64/libpthread.so.0
  2    Thread 0x45d4868461b0 (LWP 22269) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  3    Thread 0x45d48f84f1b0 (LWP 22280) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  4    Thread 0x45d4878471b0 (LWP 22270) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  5    Thread 0x45d4858451b0 (LWP 22267) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  6    Thread 0x45d4898491b0 (LWP 22272) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  7    Thread 0x45d4888481b0 (LWP 22271) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  8    Thread 0x45d48a84a1b0 (LWP 22273) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  9    Thread 0x45d48b84b1b0 (LWP 22276) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  10   Thread 0x45d48c84c1b0 (LWP 22277) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  11   Thread 0x45d48e84e1b0 (LWP 22279) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  12   Thread 0x45d48d84d1b0 (LWP 22278) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  13   Thread 0x45d4908501b0 (LWP 22281) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  14   Thread 0x45d4988681b0 (LWP 22296) 0x000045d4846f0d60 in syscall () from /lib64/libc.so.6
  15   Thread 0x45d4958651b0 (LWP 22287) 0x000045d4843ec4f0 in sigsuspend () from /lib64/libc.so.6
  16   Thread 0x45d4928521b0 (LWP 22283) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  17   Thread 0x45d4938531b0 (LWP 22284) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
  18   Thread 0x45d4968661b0 (LWP 22290) 0x000045d4846f0d60 in syscall () from /lib64/libc.so.6
  19   Thread 0x45d484843010 (LWP 22254) 0x000045d4846f0d60 in syscall () from /lib64/libc.so.6
  20   Thread 0x45d4948641b0 (LWP 22285) 0x000045d484340be0 in do_futex_wait () from /lib64/libpthread.so.0
  21   Thread 0x45d4918511b0 (LWP 22282) 0x000045d484335dc0 in pthread_cond_wait@@GLIBC_2.3.2 () from /lib64/libpthread.so.0
```

Related issue #411

**Comment by ivmai** [comment]
It is reported that a couple of bugs in signal-related code of kernel have been fixed and the issue is not observed on kernel 5.4.0-5.4.

**Comment by ivmai** [comment]
Also, probably this should fix #419

**Comment by ivmai** [comment]
> It is reported that a couple of bugs in signal-related code of kernel have been fixed and the issue is not observed on kernel 5.4.0-5.4.

I confirm the issue is not reproducible on this kernel version.
